lib/charms/data_platform_libs/v0/object_storage.py

```diff
@@ -0,0 +1,262 @@
+    """Base class for Azure storage events."""
+
+    @property
+    def container(self) -> Optional[str]:
+        """Returns the container name."""
+        if not self.relation.app:
+            return None
+
+        return self.relation.data[self.relation.app].get("container", "")
+
+
+class CredentialRequestedEvent(ContainerEvent):
+    pass
+
+
+class CredentialsChangedEvent(ContainerEvent):
+    pass
+
+
+class CredentialsGoneEvent(RelationEvent):
+    pass
+
+
+class AzureStorageProviderEvents(CharmEvents):
+    """Events for the AzureStorageProvider side implementation."""
+    credentials_requested = EventSource(CredentialRequestedEvent)
+
+
+class AzureStorageRequirerEvents(CharmEvents):
+    """Events for the AzureStorageRequirer side implementation."""
+
+    credentials_changed = EventSource(CredentialsChangedEvent)
+    credentials_gone = EventSource(CredentialsGoneEvent)
+
+
+class AzureStorageRequirerData(RequirerData):
+    SECRET_FIELDS = ["secret-key"]
+
+    def __init__(self, model, relation_name: str, container: Optional[str] = None):
+        super().__init__(
+            model,
+            relation_name,
+        )
+        self.container = container
+
+
+class AzureStorageRequirerEventHandlers(RequirerEventHandlers):
+    """Event handlers for for requirer side of Azure Storage relation."""
+
+    on = AzureStorageRequirerEvents()  # pyright: ignore[reportAssignmentType]
+
+    def __init__(
+        self, charm: CharmBase, relation_data: AzureStorageRequirerData
+    ):
+        super().__init__(charm, relation_data)
+
+        self.relation_name = relation_data.relation_name
+        self.charm = charm
+        self.local_app = self.charm.model.app
+        self.local_unit = self.charm.unit
+
+        self.framework.observe(
+            self.charm.on[self.relation_name].relation_joined, self._on_relation_joined_event
+        )
+        self.framework.observe(
+            self.charm.on[self.relation_name].relation_changed, self._on_relation_changed_event
+        )
+
+        self.framework.observe(
+            self.charm.on[self.relation_name].relation_broken,
+            self._on_relation_broken_event,
+        )
+
+    def _on_relation_joined_event(self, event: RelationJoinedEvent) -> None:
+        """Event emitted when the Azure Storage relation is joined."""
+        logger.info(f"Azure storage relation ({event.relation.name}) joined...")
+        if self.container is None:
+            self.container = f"relation-{event.relation.id}"
+        event_data = {"container": self.container}
+        self.relation_data.update_relation_data(event.relation.id, event_data)
+
+    def get_azure_connection_info(self) -> Dict[str, str]:
+        """Return the azure storage connection info as a dictionary."""
+        for relation in self.relations:
+            if relation and relation.app:
+                return self.relation_data.fetch_relation_data([relation.id])[relation.id]
```

Comment:
We should make sure that we can correctly identify the storage relation here. I.e. we should have a check if `AZURE_STORAGE_REQUIRED_INFO` was found, otherwise return `{}`. We shouldn't return potential "random fields" in this function, but either connection-related ones or none.